File: app/src/main/java/jp/hazuki/yuzubrowser/toolbar/ButtonToolbarController.java
package jp.hazuki.yuzubrowser.toolbar;

import android.graphics.ColorFilter;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.ActionCallback;
import jp.hazuki.yuzubrowser.action.manager.SoftButtonActionFile;
import jp.hazuki.yuzubrowser.settings.data.AppData;
import jp.hazuki.yuzubrowser.utils.view.swipebutton.SwipeImageButton;

public class ButtonToolbarController {
    private final int TOOLBAR_SIZE_Y;
    private final ViewGroup linearLayout;
    private final ActionCallback mActionCallback;
    private SwipeImageButton mButtonList[];

    public ButtonToolbarController(ViewGroup layout, ActionCallback action_callback, int toolbar_size) {
        linearLayout = layout;
        mActionCallback = action_callback;
        TOOLBAR_SIZE_Y = toolbar_size;
    }

    public void addButtons(List<SoftButtonActionFile> list) {
        int size = list.size();
        if (mButtonList != null) {// initialized
            if (mButtonList.length == size) {
                int i = 0;
                for (SwipeImageButton btn : mButtonList) {
                    btn.setActionData(list.get(i), mActionCallback);
                    btn.setSense(AppData.swipebtn_sensitivity.get());
                    ++i;
                }
            } else {
                linearLayout.removeAllViews();

                mButtonList = new SwipeImageButton[size];
                LayoutInflater inflater = LayoutInflater.from(linearLayout.getContext());

                for (int i = 0; i < size; ++i) {
                    SwipeImageButton btn = inflateButtonView(inflater, linearLayout);
                    mButtonList[i] = btn;
                    btn.setActionData(list.get(i), mActionCallback);
                    btn.setSense(AppData.swipebtn_sensitivity.get());
                    settingButtonSize(btn, TOOLBAR_SIZE_Y);
                }
            }
        } else {
            mButtonList = new SwipeImageButton[size];
            LayoutInflater inflater = LayoutInflater.from(linearLayout.getContext());

            for (int i = 0; i < size; ++i) {
                SwipeImageButton btn = inflateButtonView(inflater, linearLayout);
                mButtonList[i] = btn;
                settingButtonSize(btn, TOOLBAR_SIZE_Y);
            }
        }
    }

    public SwipeImageButton inflateButtonView(LayoutInflater inflater, ViewGroup parent) {
        View view = inflater.inflate(R.layout.toolbar_custom_button, null);
        parent.addView(view);
        return (SwipeImageButton) view.findViewById(R.id.button);
    }

    public static void settingButtonSize(View view, int size) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        params.height = size;
        params.width = size;
        view.setLayoutParams(params);
    }

    public static void settingButtonSizeHeight(View view, int size) {
        ViewGroup.LayoutParams params = view.getLayoutParams();
        params.height = size;
        view.setLayoutParams(params);
    }

    public void notifyChangeState() {
        for (SwipeImageButton btn : mButtonList) {
            btn.notifyChangeState();
        }
    }

    public void setColorFilter(ColorFilter cf) {
        for (SwipeImageButton btn : mButtonList)
            btn.setColorFilter(cf);
    }

    public void setBackgroundDrawable(Drawable background) {
        for (SwipeImageButton btn : mButtonList)
            btn.setImageDrawable(background);
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/toolbar/ButtonToolbarController.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.toolbar

import android.graphics.ColorFilter
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import jp.hazuki.yuzubrowser.legacy.R
import jp.hazuki.yuzubrowser.legacy.action.manager.ActionController
import jp.hazuki.yuzubrowser.legacy.action.manager.ActionIconManager
import jp.hazuki.yuzubrowser.legacy.action.manager.SoftButtonActionFile
import jp.hazuki.yuzubrowser.legacy.utils.view.swipebutton.SwipeImageButton
import jp.hazuki.yuzubrowser.ui.settings.AppPrefs

open class ButtonToolbarController(private val linearLayout: ViewGroup, private val controller: ActionController, private val iconManager: ActionIconManager, private val TOOLBAR_SIZE_Y: Int) {
    private val mButtonList: ArrayList<SwipeImageButton> = ArrayList(1)

    fun addButtons(list: List<SoftButtonActionFile>) {
        val size = list.size
        if (mButtonList.size == size) {
            for ((i, btn) in mButtonList.withIndex()) {
                btn.setActionData(list[i], controller, iconManager)
                btn.setSense(AppPrefs.swipebtn_sensitivity.get())
            }
        } else {
            linearLayout.removeAllViews()

            val inflater = LayoutInflater.from(linearLayout.context)

            for (i in 0 until size) {
                val btn = inflateButtonView(inflater, linearLayout)
                mButtonList.add(btn)
                btn.setActionData(list[i], controller, iconManager)
                btn.setSense(AppPrefs.swipebtn_sensitivity.get())
                settingButtonSize(btn, TOOLBAR_SIZE_Y)
            }
        }
    }

    open fun inflateButtonView(inflater: LayoutInflater, parent: ViewGroup): SwipeImageButton {
        val view = inflater.inflate(R.layout.toolbar_custom_button, parent, false)
        parent.addView(view)
        return view.findViewById(R.id.button)
    }

    fun notifyChangeState() {
        for (btn in mButtonList) {
            btn.notifyChangeState()
        }
    }

    fun resetIcon() {
        for (btn in mButtonList) {
            btn.setToDefault()
        }
    }

    fun setColorFilter(cf: ColorFilter?) {
        for (btn in mButtonList)
            btn.colorFilter = cf
    }

    fun setBackgroundDrawable(background: Drawable?) {
        for (btn in mButtonList)
            btn.setImageDrawable(background)
    }

    companion object {

        fun settingButtonSize(view: View, size: Int) {
            val params = view.layoutParams
            params.height = size
            params.width = size
            view.layoutParams = params
        }

        fun settingButtonSizeHeight(view: View, size: Int) {
            val params = view.layoutParams
            params.height = size
            view.layoutParams = params
        }
    }
}
